### PR TITLE
Allow usage of Zapatos from within node_modules

### PR DESCRIPTION
**Pull request by eyelidlessness**
In the discussion in #16, this is a working approach to allowing Zapatos to be used within node_modules, no copying or symlinking necessary. Some notes on the changes and the approach:

- For Zapatos users (via `npm install zapatos`), this *shouldn't* require a tsconfig change, so long as the generated schema file ends up in a path the project looks for TS files.
- For the Zapatos development, there *is* a `paths` config, it basically mirrors the implicit config for `node_modules`, to allow the internal `schema.d.ts` to be picked up as `zapatos/schema`. This allows replacing the relative path imports with a stable module path `zapatos/schema`. TypeScript does not include this internal `schema.d.ts` in the build, so there will be no ambiguity when users generate their own schema.
- I've removed the copy and symlink code paths in this PR. I understand there may still be use cases for those configurations (the most obvious one which comes to mind, as mentioned in #16, is projects which access multiple databases), but mainly I wanted to keep this simple for easy review.
- A change to `package.json` was necessary: `main` now references `src` rather than `generate`. This is almost certainly a breaking change to the programmatic generation APIs, which I want to call out. I didn't address this mainly because I wanted to make sure to get this PR open before the weekend is out, and I don't feel familiar enough in the codebase yet to attempt to resolve it with confidence.

I manually verified that this worked on a simple local dummy project with the following steps:

1. In Zapatos: `npm run build`. I believe this step can be skipped if you rename `preversion` to `prepack`, but I didn't want to be presumptuous about the build/publish flow.
2. In Zapatos: `npm pack`. This produces a tarball of the build similar to what would be published to NPM.
3. In test database: `CREATE TABLE "TestZapatosNodeModules" ("name" TEXT NOT NULL);`
4. In test project: `npm install ../zapatos/zapatos-0.1.51.tgz`.
5. In test project: `./node_modules/.bin/zapatos`
6. In test project:

    ```typescript
    import * as pg from 'pg';
    import * as db from 'zapatos';
    import * as s  from 'zapatos/schema';

    type Equals<X, Y> =(<T>() => T extends X ? 1 : 2) extends (<T>() => T extends Y ? 1 : 2) ? true : false;
    type PromiseType<T extends Promise<any>> = T extends Promise<infer U> ? U : never;

    const pool = new pg.Pool({});

    export const testSQL = async () => {
        const [ { name } ] = await db.sql<s.TestZapatosNodeModules.SQL, s.TestZapatosNodeModules.Selectable[]>`
            SELECT * FROM ${"TestZapatosNodeModules"}
        `.run(pool);

        return name;
    };

    const testSQLSuccess: Equals<PromiseType<ReturnType<typeof testSQL>>, string> = true;

    export const testShortcut = async () => {
        const [ { name } ] = await db.select('TestZapatosNodeModules', db.all).run(pool);

        return name;
    };

    const testShortcutSuccess: Equals<PromiseType<ReturnType<typeof testShortcut>>, string> = true;
    ```

### CODE REVIEW

**Comment by eyelidlessness** [comment]
Oh, I forgot to mention:

- One small caveat, which I could look into if it's important, is there's a bit of an oddity when navigating into `core.d.ts` and `shortcuts.d.ts` within `node_modules`, where VSCode appears to forget that it knows where `zapatos/schema` is. I don't know how much of a problem this might cause for DX, but wanted to be transparent about it as a known issue.
- I'm sure if this change (or one like it) is accepted, there will of course need to be updates to docs. I'm not really sure how to go about that but I'd be happy to take a look.

**Comment by jawj** [comment]
Thanks, this looks very promising! I just converted my own zapatos-demo repo, where I try things out, to this way of doing things, and it wasn't too painful (I run the main script here with `ts-node`, so I had to add a `files` directive to `tsconfig.json`, and also the `--files` flag to the `ts-node` shebang, but then it worked perfectly).

This is quite a major change, so give me a little while to reflect. I'll also come back to you in more detail on the points you raise in the next few days.

**Comment by eyelidlessness** [comment]
I'm glad it wasn't too painful in your demo repo, that is promising! I didn't think to try with `ts-node` for validation in this case because I didn't expect a difference, but now I recall having to set `--files` in projects where I use it too, so that makes sense and I'll keep that in mind.

No rush at all on this of course, I'm sure I won't be able to do much on this during the week anyhow.

**Comment by jawj** [comment]
Update: I'm now trying to convert the larger (proprietary) project that Zapatos emerged from to no-copy, and having trouble with both ts-jest and ts-node not finding the ambient `'zapatos/schema'` type declaration.

I'll try some more to get these working, but I do worry that these sorts of issues might cause a lot of frustration and cause people to give up ...

**Comment by jawj** [comment]
Sorry for the slow progress here. I think what I probably need to do next is extract a subset of the proprietary project that I can share.

**Comment by jawj** [comment]
Hmm — 2 months later, and I just can't seem to bring myself to get into these issues with ts-jest and ts-node. I think it's because I know from experience that these tools can be very opaque and frustrating, and also because I guess that if I (as project maintainer) find it hard to gear up to fix this, what chance is there that prospective library users would be willing to?

However, I'm still very keen that we get some tests started for the project: as I add new features, it's always a bit nail-biting not to have proper test coverage to ensure no old features get broken in the process.

So, @eyelidlessness: would you be willing to park this issue for now, and still take a look at the testing side when you have some free time?

**Comment by jtfell** [comment]
Hey @jawj - I wanted to jump in here and say that this issue is the only thing stopping me from trying out Zapatos. I can't justify the extra complexity of sym-linking to my team, but really love the overall approach of the library. I'd be happy with some ts configuration tweaks rather than a custom build-step that might not be portable. Just my 2c :)

**Comment by jawj** [comment]
@jtfell Thanks for your comments here.

- For clarity, the library's default approach is not to symlink, but to copy the relevant source files to a `src` folder alongside the generated `schema.ts` file, inside the `zapatos` folder which lives somewhere in your source tree.
- However we approach the source files, I think we'll always have to generate `schema.ts` based on your database schema, and dump it somewhere in your source tree. And I'm guessing that this will always require a 'custom build-step' (though really it needs to be done after `npm update` rather than at build time, because how you write the rest of your code depends on it).
- One advantage of copying the Zapatos source files is that we should always end up with a schema and source files that are mutually compatible. If the source files stayed in `node_modules` but referenced `schema.ts` within your source, then some Zapatos updates would leave you with source files that didn't play nice with your schema until you regenerated it.
- My previous experience is that I have simply not been able to make either `ts-jest` or `ts-node` compile TypeScript files anywhere inside `node_modules`, however much I tweak their options.

Given all the above, my overall feeling is: we're always going to need a build step, and we're always going to have to drop at least _some_ of our code into your code. Therefore, isn't it better to do it the (slightly inelegant) way that just works, and drop the whole lot in, rather than the (slightly neater) way that tends to break if you do anything beyond running plain old `tsc` on your project?

Note that this is a genuine question: please do tell me what you think. I find this approach to DB access really handy, and I'm keen to make it possible for more people to get on board with it.

I'm also not quite sure what you mean when you say the build step "might not be portable"?

**Comment by jtfell** [comment]
Thanks for the thoughtful response. It's definitely a tricky situation, and I disagree on some of the points but please don't read too much into my opinion on this. I've not spent a huge amount of time thinking about it and its just my initial impressions I'm going on. 

A few thoughts:

- Vendoring in dependencies to your codebase is often seen as a smell and pollutes git diffs. As the `schema.ts` file is generated based on the project data, it does not fit this category so does belong in the repo in my opinion.
- Users will be less likely to update to the latest version if they have to do it manually. 
- It's a good point you make about the generated schema and library getting out of sync. I suspect you would be able to detect this mismatch at runtime and log an error, but I may be missing some detail here.
- Why do you need to compile TypeScript files inside node_modules? Isn't the usual pattern for the published package to contain pre-compiled JS and generated typescript interfaces? Or is there a technical reason TS files are needed in the library?
- How do other similar projects achieve this kind of thing? Eg. [Prisma](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client#set-up) goes the other way and even puts the generated schema back into node_modules which brings its own issues with it..

More generally, I think that a more clunky configuration (or code initialisation) is preferable to requiring users to copy the source into the repo. 

**Note:** With "not portable" I mean't symlinking is different on different OSes but this seems to be less of a problem if you copy the source.

**Comment by cpgo** [comment]
> Why do you need to compile TypeScript files inside node_modules? Isn't the usual pattern for the published package to contain pre-compiled JS and generated typescript interfaces? Or is there a technical reason TS files are needed in the library?

I belive we cant have a pre-compiled `schema.ts` because it is generated based on the database schema.

In the end, I don't see much of a benefit on "hiding" the generated schema on `node_modules`.

> Users will be less likely to update to the latest version if they have to do it manually.

We kinda already have to build our schema.ts every time we change our database schema, I think it is reasonable to expect the developer to run zapatos again after updating the package.
Or we could try and make it less manual by creating some `postmigration`/`prebuild` scripts on our package.json

**Comment by eyelidlessness** [comment]
Brief head above water: I'm incredibly regretting trying to make a big change for a library I hold in very high regard and dropping off the planet not once but twice.

I went through two burnout waves, one during my last job and the next one following my resignation. I've been wanting to come back and contribute to this, and if there's space for the contribution when I'm ready I hope I can dig deep into this, because it's exactly my cup of tea.

But I'm taking the responsibility right now to say I'm sorry I've avoided it and gone dark for so long. That's really not great footing to start off on. I have some personal projects on my horizon that would benefit from making strides with zapatos, and I have every intention of coming back to see how I can support it when I'm not in "briefly came up for air" mode.

**Comment by jawj** [comment]
@jtfell I'll continue the conversation in your new issue, but it was easier to reply on some of these points here:

>     * Vendoring in dependencies to your codebase is often seen as a smell and pollutes git diffs. As the `schema.ts` file is generated based on the project data, it does not fit this category so does belong in the repo in my opinion.

OK, the diffs issue I'm somewhat persuaded by.

>     * Users will be less likely to update to the latest version if they have to do it manually.

Kind of unavoidable, given the need to regenerate `schema.ts`?

>     * It's a good point you make about the generated schema and library getting out of sync. I suspect you would be able to detect this mismatch at runtime and log an error, but I may be missing some detail here.

Fair point, we can probably do that.

>     * Why do you need to compile TypeScript files inside node_modules? Isn't the usual pattern for the published package to contain pre-compiled JS and generated typescript interfaces? Or is there a technical reason TS files are needed in the library?

Hmm, it's possible I misdescribed this.

> More generally, I think that a more clunky configuration (or code initialisation) is preferable to requiring users to copy the source into the repo.

As long as it's straightforward to get it working, then I guess so — though perhaps more as removing a barrier to adoption than for actual DX.

Anyway, in summary, I'll take another look at the problems I had with this before.

**Comment by jawj** [comment]
@eyelidlessness No worries: take care of yourself. Look forward to having you back if and when you have the space.

**Comment by jawj** [comment]
Closing this PR because the same work is ongoing (and almost completed) in #44.
